### DISCUSSION TITLE
blog/2022/221113-rclone-sharepoint-for-hentai-at-home/

### DISCUSSION THREAD

**Discussion by giscus** [zhanbao2000/blog#17] [Announcements]
# blog/2022/221113-rclone-sharepoint-for-hentai-at-home/

在服务器上挂 H@H 时，总会遇到空间不足的情况，这篇指南介绍了使用 Rclone 挂载 SharePoint 作为 Hentai@Home 的缓存目录的方法，以极大地扩展我们 H@H 的缓存空间。

https://blog.arisa.moe/blog/2022/221113-rclone-sharepoint-for-hentai-at-home/

**Comment by 1071974393** [comment]
你好博主，我想问问通过这种方法把onedrive挂载到vps并且把H@H缓存目录放入挂载目录中，这样能提升H@H缓存的大小吗，H@H的缓存文件不是仍然存在于vps的本地硬盘中吗，缓存容量的瓶颈不还是在vps的硬盘上，我不是很理解这种方式对于H@H的意义体现在哪里，能帮忙解释一下吗

**Reply by zhanbao2000** [reply]
这种方式的意义其实就是提升 H@H 的缓存大小。

**rclone 把 OneDrive 挂载到 vps 上的一个目录，并且把 H@H 缓存目录指定为这个目录，此后 H@H 对于这个目录的读写操作，最终都会变成对 OneDrive 的读写操作。H@H 缓存容量取决于你的 OneDrive 容量大小**。

另外 H@H 的缓存文件不存放在 vps 本地硬盘中。rclone 在挂载的时候提供了一个虚拟文件系统，对该文件系统进行读写会被 rclone 接管，rclone 在远程进行等价的读写操作。

至于你说的 vps 的本地磁盘，它唯一的意义在于 rclone 在本地和远程之间通信时充当缓存，从而降低网络请求的频率，它是真正意义上的缓存，和 H@H 无关，它的目录在 `/root/.cache/rclone/vfs`，它的大小可以由 `--vfs-cache-max-size` 参数控制，它不会影响 H@H 的缓存大小。

例如我的 OneDrive 空间为 1TB，我的 vps 磁盘剩余空间为 10GB。那么我通过此博客介绍的方法，挂载 OneDrive 到 vps 并将其指定为 H@H 缓存目录后，我可以在 e-hentai 上将该 H@H 实例的 `Maximum Disk Cache Size` 值设置为 900GB。但是如果我直接把 vps 磁盘作为 H@H 缓存目录，那么我只能将该值设置为 9 GB（注意当磁盘可用空间低于 1 GB 时，H@H 客户端会强制退出）。

**Reply by 1071974393** [reply]
明白了，感谢回复

**Comment by Volundio** [comment]
想请问下博主，你h@h挂Onedrive后，E站上显示的Quality值是多少？我拿google drive挂，质量已经掉的马上不到3k了。。。2k客户端就要被停用了。。。
Trust倒是一直是1k满的
看了下h@h客户端的log，也没报啥错

**Reply by zhanbao2000** [reply]
最开始挂的时候变成多少我已经不记得了，最后稳定下来的时候大概8000左右

**Reply by Volundio** [reply]
> 最开始挂的时候变成多少我已经不记得了，最后稳定下来的时候大概8000左右

请问一下你是拿官方java客户端挂的吗？有跑docker挂吗？

**Reply by zhanbao2000** [reply]
就是拿官方给的java挂的，没用docker

**Reply by Volundio** [reply]
> 就是拿官方给的java挂的，没用docker

我还以为是我用docker跑的rust重构的第三方客户端  https://github.com/james58899/hath-rust  有问题。。

这会换回官方客户端，好像还在掉，麻了。。

想请问一下你的rclone参数是有针对优化过吗？
--allow-other --transfers 4 --buffer-size 32M --vfs-read-chunk-size 128M --vfs-read-chunk-size-limit 0M --vfs-cache-max-size 8000M --vfs-cache-max-age 3360h --allow-non-empty --dir-cache-time 3360h --vfs-cache-mode full --umask 000 --vfs-cache-mode full --no-modtime --vfs-cache-poll-interval 30m


编辑，改成用你这份参数也还在掉Quality。。现在真的连2000都没了，Trust也还是1k满的。。
难道是gd不如onedrive？

**Reply by zhanbao2000** [reply]
> > 就是拿官方给的java挂的，没用docker
> 
> 我还以为是我用docker跑的rust重构的第三方客户端 https://github.com/james58899/hath-rust 有问题。。
> 
> 这会换回官方客户端，好像还在掉，麻了。。
> 
> 想请问一下你的rclone参数是有针对优化过吗？ --allow-other --transfers 4 --buffer-size 32M --vfs-read-chunk-size 128M --vfs-read-chunk-size-limit 0M --vfs-cache-max-size 8000M --vfs-cache-max-age 3360h --allow-non-empty --dir-cache-time 3360h --vfs-cache-mode full --umask 000 --vfs-cache-mode full --no-modtime --vfs-cache-poll-interval 30m
> 
> 编辑，改成用你这份参数也还在掉Quality。。现在真的连2000都没了，Trust也还是1k满的。。 难道是gd不如onedrive？

和文章中用的一样的参数

确实很奇怪，按理说如果rclone能正常提供文件，然后H@H也没有出现大量报错（理论上正常请求到你的H@H都是 200 OK），那这个Quality应该会缓慢上升才对。我个人觉得和gd没有关系，可能是因为重置了静态范围导致的掉Quality，如果真是这样的话只要保持Trust值正常，Quality到最后都会升回来

**Reply by Volundio** [reply]
> > > 就是拿官方给的java挂的，没用docker
> > 
> > 
> > 我还以为是我用docker跑的rust重构的第三方客户端 https://github.com/james58899/hath-rust 有问题。。
> > 这会换回官方客户端，好像还在掉，麻了。。
> > 想请问一下你的rclone参数是有针对优化过吗？ --allow-other --transfers 4 --buffer-size 32M --vfs-read-chunk-size 128M --vfs-read-chunk-size-limit 0M --vfs-cache-max-size 8000M --vfs-cache-max-age 3360h --allow-non-empty --dir-cache-time 3360h --vfs-cache-mode full --umask 000 --vfs-cache-mode full --no-modtime --vfs-cache-poll-interval 30m
> > 编辑，改成用你这份参数也还在掉Quality。。现在真的连2000都没了，Trust也还是1k满的。。 难道是gd不如onedrive？
> 
> 和文章中用的一样的参数
> 
> 确实很奇怪，按理说如果rclone能正常提供文件，然后H@H也没有出现大量报错（理论上正常请求到你的H@H都是 200 OK），那这个Quality应该会缓慢上升才对。我个人觉得和gd没有关系，可能是因为重置了静态范围导致的掉Quality，如果真是这样的话只要保持Trust值正常，Quality到最后都会升回来

这次用GD挂之前确实把静态范围重置了。
挂了几天，现在Quality从不到2000回升到4000浮动，但是也没法再上去了。。。log里也没报啥错误

**Reply by zhanbao2000** [reply]
我的建议是再静候一段时间，毕竟quality的变化确实需要很长的周期

> Quality measures the long-term overall stability and reliability of a client, ...

如果过了相当长的时间还是卡在4000，那可能真的就是gd的上限了

**Comment by TransparentLC** [comment]
想问一下这套方案对配置有要求吗？或者有什么可以降低内存占用的建议吗？

我有个单核 1GB 内存的 VPS，开着 3GB 的 zram 作为 swap，因为原版 hath 内存占用太多所以用的是 hath-rust，按照这套方案用 rclone 挂载后内存占用经常窜到最大 350MB 然后触发 swap 占满 CPU，严重的时候服务器甚至会直接死机，于是现在进入了 quality 归零→hath 不处理请求→quality 上升到 2000→hath 开始处理请求且不断增多→服务器卡死→quality 归零的循环，这样下来 quality 别说 4000 了达到 3000 都成问题。

现在的挂载命令如下：`rclone mount onedrive: /mnt/onedrive --use-mmap --allow-other --vfs-cache-mode full --vfs-cache-max-size 1g --vfs-cache-max-age 24h --transfers 4 --buffer-size 512k --attr-timeout 1m --user-agent "ISV|rclone.org|rclone/v1.72.0" --verbose`

**Reply by zhanbao2000** [reply]
具体配置要求我没研究过，我现在是1G内存+2Gswap。这里面的占用大头就是rclone，我这正常挂几天能用到我500多MB，rclone的内存占用就是这样的，没办法

建议的降低内存占用率的方法，也就是我文章里说的那些，比如 --use-mmap 参数，--buffer-size 参数。我看你在命令行里都已经有写了

不过一般来说只是单纯触发swap不会把CPU占满的，我只遇到过内存和swap双双占满，然后才导致cpu被kswapd0占满的情况

**Reply by TransparentLC** [reply]
> 不过一般来说只是单纯触发swap不会把CPU占满的

zram 是拿内存中的一部分作为 swap，但是加了一层压缩，这样的话也许是压缩和解压缩占了 CPU？现在设的压缩算法是 zstd，打算换成压缩率低一些但是速度快一些的 lz4 再观察一段时间，实在不行就改回使用普通的 swapfile。

**Reply by zhanbao2000** [reply]
哦不好意思没注意到你用的是zram，我在我vps上用的是swapfile。这么说确实有可能是压缩解压导致的CPU占用

**Reply by TransparentLC** [reply]
试过了使用普通的 swapfile 并且不用 zswap 之类的压缩技术，不过就算这样 quality 也上不了 3000，翻了下日志发现上 2000 之后大部分请求都能在 1000ms 内响应完但是到后面经常出现几个响应时间高达数十甚至上百秒的请求最后 502，同时监控的邮件提醒也告诉我这个时候 hath 已经卡死了。

然后我写了个脚本想模拟一下 hath 是如何读取 cache 的。一张 cache 的图片的路径形如 `cache/aa/bb/aabbccddeeff...`，于是脚本里先从 `cache` 随机取出 `aa`，再从 `cache/aa` 随机取出 `bb`，再从 `cache/aa/bb` 随机取一个文件完整读取，同时开八个协程循环做这个操作：

```py
import aiofiles
import aiofiles.os
import asyncio
import random
import os
import time

CACHE_DIR = '/mnt/onedrive/hath/cache'

listCount = 0
listTime = 0
readCount = 0
readTime = 0

lock = asyncio.Lock()

async def worker():
    global listCount, listTime, readCount, readTime
    while True:
        try:
            ts = time.perf_counter()
            depth1 = os.path.join(CACHE_DIR, random.choice(await aiofiles.os.listdir(CACHE_DIR)))
            depth2 = os.path.join(depth1, random.choice(await aiofiles.os.listdir(depth1)))
            depth3 = os.path.join(depth2, random.choice(await aiofiles.os.listdir(depth2)))
            te = time.perf_counter()
            t = te - ts
            listCount += 1
            listTime += t
            print('Listing', listCount, (listTime / listCount) if listCount else 0, 'Reading', readCount, (readTime / readCount) if readCount else 0)

            ts = time.perf_counter()
            async with aiofiles.open(depth3, 'rb') as f:
                d = await f.read()
            te = time.perf_counter()
            t = te - ts
            readCount += 1
            readTime += t
            print('Listing', listCount, (listTime / listCount) if listCount else 0, 'Reading', readCount, (readTime / readCount) if readCount else 0)

            await asyncio.sleep(random.randint(0, 3000) / 1000)
        except Exception as ex:
            print(ex)

async def main():
    await asyncio.gather(*(worker() for _ in range(8)))

if __name__ == '__main__':
    asyncio.run(main())
```

结果三个 `os.listdir` 平均就得几十秒，读取文件也得平均几十秒，更诡异的是测到后面 VPS 直接卡死了，btop 里找不到大量占用 CPU 的进程但是负载又跑到 5 以上了。如果是用之前硬盘上的 `cache` 平均只需要 0.5s 和 0.1s。我还按照文档在挂载参数里加了 `--onedrive-delta --fast-list`，不过也没有效果。

总之我基本打算放弃用 rclone 挂载了，等 cache 转移回来老老实实继续用硬盘挂 hath 吧。留个反面案例在这~~引以为戒~~，这再折腾下去费时费力，而且 quality 如此糟糕别说扩大 static range 了菠萝不把我的机拆掉都算是手下留情了 (╯‵□′)╯︵┻━┻

我的 VPS 是 Vultr 东京的，不知道会不会存在线路问题访问 onedrive 不畅。之前把硬盘上的 cache 通过 rclone sync 同步到 onedrive 的时候也经常出现所有上传突然卡住的情况，一个小时可能只能上传 500MB 左右，不过那种时候没有出现大量占用 CPU 或内存或者卡死的情况。

**Reply by TransparentLC** [reply]
卡死可能是因为 onedrive 限流了，加了 `-vv` 之后看到 sync 和 mount 的日志里有大量 `DEBUG : pacer: Rate limited, increasing sleep to 1m6s`、`DEBUG : pacer: Reducing sleep to 12.614227294s`、`INFO  : OneDrive root 'rclone': Change notify listener failure: activityLimitReached: throttledRequest: The request has been throttled` 这样的记录。

根据文档说的我有在命令里加上[那个 ISV 的 User-Agent](https://rclone.org/onedrive/#excessive-throttling-or-blocked-on-sharepoint) 以及[使用自己单独注册的 Client ID/Secret](https://rclone.org/onedrive/#getting-your-own-client-id-and-key)，不过这样看来也没什么效果……
